Railsでユーザーフォローを作る

pofkuma opened the pull request:
「Railsでユーザーフォローを作る」の提出物を作成しました。

ご確認よろしくお願いします。

pofkuma committed:
ユーザーフォローの関連設定

フォローする／されるの関連を作成する

pofkuma committed:
フォローボタンの設置

ユーザー詳細画面にアクセスしてそのユーザーをフォローできるようにする

pofkuma committed:
Followの関連をユニークにする

「フォロワーに全く同じユーザーが2人いる」ということが起きないように、インデックスとバリデーションを追加

pofkuma committed:
バリデーションエラーを表示させる

「フォローする」のHTTPリクエストを再送した時にエラーメッセージを表示できるようにする

pofkuma committed:
フォロー解除ボタンの設置

すでにフォローしてる場合はフォロー解除できるようにする

pofkuma committed:
フォロー数とフォロワー数の表示

ユーザー詳細画面に現在のフォロー数（そのユーザーがフォローしているユーザーの人数）とフォロワー数を表示する

pofkuma committed:
翻訳階層の見直し

lazy lookup を利用する

pofkuma committed:
フォロー一覧とフォロワー一覧のページ追加

ユーザー詳細画面から、フォロー一覧画面とフォロワー一覧画面に遷移できるようにする

pofkuma committed:
フォロワーの関連の命名を統一する

フォローはfollowingで統一しているので、フォロワーはfollowerで統一する

pofkuma committed:
フォロー一覧とフォロワー一覧にページングを追加

pofkuma commented on db/schema.rb:
```diff
@@ -35,7 +35,7 @@
 
   create_table "active_storage_variant_records", force: :cascade do |t|
-    t.bigint "blob_id", null: false
+    t.integer "blob_id", null: false
```
特に手は加えていないと思うのですが、データ型が変わっていました。
直近の変更を見ると元々は`integer`だったようですが…。
https://github.com/fjordllc/fjord-books_app/blob/08908a1ccefec1accd27b1f69759c5ec4d317c70/db/schema.rb#L38
https://github.com/pofkuma/fjord-books_app/commits/07-follow_users/db/schema.rb

betachelsea replied:
こちら大丈夫です！ 🙆  確認コメントありがとうございます！

betachelsea commented on app/controllers/follows_controller.rb:
```diff
@@ -0,0 +1,29 @@
+  def create
+    follow = current_user.following_relationships.build(following_id: params[:user_id])
+
+    respond_to do |format|
```
[MUST] レスポンスの形式が変わらない場合、`respond_to` は使用なしで大丈夫です。
使い所としては、 htmlを求められたらhtmlを返したいけど、jsonでの応答を指定されたらjsonで返事をしたい...といった場合に利用します。

```ruby
# 参考: Rails 6.1.3.2 (ちょっと古くてすみません) scaffoldの実行で生成したソースより

# POST /items or /items.json
def create
  @item = Item.new(item_params)

  respond_to do |format|
    if @item.save
      format.html { redirect_to @item, notice: "Item was successfully created." }
      format.json { render :show, status: :created, location: @item }
    else
      format.html { render :new, status: :unprocessable_entity }
      format.json { render json: @item.errors, status: :unprocessable_entity }
    end
  end
end
```

↓htmlだけでいい場合

```ruby
def create
  @item = Item.new(item_params)

  if @item.save
    redirect_to @item, notice: "Item was successfully created."
  else
    render :new, status: :unprocessable_entity
  end
end
```

pofkuma replied:
そうなのですね、ありがとうございます！
htmlのみの想定でも`responf_to`を使うものだと思い込んでおりました…。
de5d731ad146e72443e45b1fd360cd8bd409a94c にて修正しましたので、ご確認をお願いします🙏

betachelsea commented on app/controllers/follows_controller.rb:
```diff
@@ -0,0 +1,29 @@
+      elsif follow.errors.where(:following).any?
+        format.html { redirect_to user_path(params[:user_id]), alert: follow.errors.where(:following).last.full_message }
+      else
+        format.html { redirect_to user_path(params[:user_id]) }
```
[NITS] こちらについて、

- followingだけ、など個別のエラーを切り出す実装をしているとどんどんソースが肥大化してしまう
- 最後のelse分岐に辿り着いた場合、ユーザは何が発生したのかわからない(メッセージがないため)

という問題がありそうだなと思いました。エラーの区別なく、`follow` の抱えているエラーメッセージ全部出してしまう or エラーがあったということだけが肝要ならば、固定メッセージを出すとかで対処して分岐を減らしても良いかなと思いました 👀

pofkuma replied:
仰る通りだと思いました…！
今回は、エラーがあったことだけがわかるように固定メッセージを表示させるように修正しました。
`destroy`の方も併せて修正しました。 -> 6bf23413aad7df80d156cd35851b2946e2c7d659

betachelsea commented on app/controllers/follows_controller.rb:
```diff
@@ -0,0 +1,29 @@
+      if follow.destroy
+        format.html { redirect_to user_path(params[:user_id]), notice: t('.notice') }
+      else
+        format.html { redirect_to user_path(params[:user_id]) }
```
[NITS] エラーの時にユーザーへの何かしら案内があるといいなと思いました 👀

pofkuma replied:
`create`の方と併せて修正いたしました🙏
-> 6bf2341

betachelsea commented on app/views/users/followers/index.html.erb:
```diff
@@ -0,0 +1,33 @@
+<h1><%= t('.title') %></h1>
+
+<table>
```
[IMO] `app/views/users/followings/index.html.erb` と中身がほぼ同じなので、うまくpartial機能をつかって共通化してみても良さそうですね。

[レイアウトとレンダリング \- Railsガイド - 3.4.3 パーシャルレイアウト](https://railsguides.jp/layouts_and_rendering.html#%E3%83%91%E3%83%BC%E3%82%B7%E3%83%A3%E3%83%AB%E3%83%AC%E3%82%A4%E3%82%A2%E3%82%A6%E3%83%88)

pofkuma replied:
確かに共通化した方が良さそうです、ありがとうございます！
8facc94126a8761a611d637ac3fe4857fe2447f5 にて対応してみました。

betachelsea commented on app/views/users/followers/index.html.erb:
```diff
@@ -0,0 +1,33 @@
+
+<%= paginate @followers %>
+
+<%= link_to t('views.common.back'), user_path(params[:user_id]) %>
```
[NITS] `user_path` に渡す値は `params[:user_id]` そのままではなく、`User.find(params[:user_id])` した結果(Userのインスタンス)であれば、確実に遷移先にユーザが存在する保証ができそうですね。

pofkuma replied:
コントローラーで生成したインスタンスを利用すると良いという認識で、 72ff36887afad4161adb766b49fce81e7fd1155c のように修正してみました。
もし間違っていれば教えてください🙇‍♀️

betachelsea commented on app/views/users/show.html.erb:
```diff
@@ -33,6 +33,28 @@
+
+<% unless current_user == @user %>
+  <p>
+    <% if current_user.following_relationships.find_by(following_id: @user.id) %>
```
[NITS]存在有無を確認したいときは、exists? も使えます。

```ruby
current_user.following_relationships.where(following_id: @user.id).exists?
```

betachelsea replied:
[IMO] 使い回しの効きそうな判定処理がある場合、Model側にメソッドを生やしてしまうのもありですね〜。

pofkuma replied:
e3f171fb737a60aac6f62f326422bfae42c41c1c にてメソッド化し、`exists?`を使うようにしました🙏

betachelsea commented on app/views/users/show.html.erb:
```diff
@@ -33,6 +33,28 @@
+<hr>
+
+<p>
+  <strong>フォロー数:</strong> <%= @user.following_relationships.count %> |
+  <strong>フォロワー数:</strong> <%= @user.follower_relationships.count %>
```
[NITS] I18nを徹底しようと思うと、このあたりにも適用したいですね！（今回の必須要件には入っていなさそう？なので、このままでもOKとします 🙏  )

pofkuma replied:
うっかり直接書いてしまっていました💦
982912de9c06842c517c6e60764660b72ebccc47 にて修正しました！

betachelsea commented on db/migrate/20220116093317_add_index_to_follow.rb:
```diff
@@ -0,0 +1,7 @@
+
+class AddIndexToFollow < ActiveRecord::Migration[6.1]
+  def change
+    add_index(:follows, %i[following_id follower_id], unique: true, name: 'by_following_follower')
```
[NITS] `add_index` は create_table の時のmigrationファイルと一緒にしておくこともできますよ！

```ruby
# frozen_string_literal: true

class CreateFollows < ActiveRecord::Migration[6.1]
  def change
    create_table :follows do |t|
      t.references :following, null: false, foreign_key: { to_table: :users }
      t.references :follower, null: false, foreign_key: { to_table: :users }

      t.timestamps
    end
    # ここに書く
    add_index(:follows, %i[following_id follower_id], unique: true, name: 'by_following_follower')
  end
end

```

pofkuma replied:
create_table の所に一緒に書けるのですね💡
ユニーク制約が必要だと後から気づいて追加しておりました。
教えてくださりありがとうございます、今後覚えておきたいと思います！

betachelsea reviewed (commented):
[MUST] 扱いとした `respond_to` を使わない書き方について、ぜひ慣れていただきたく、その点は必須で修正してもらいたいです。 残りの細かい指摘についての修正有無はおまかせします。よろしくお願いします！ 🙏

pofkuma committed:
htmlでのレスポンスのみのため respond_to は不要

https://github.com/pofkuma/fjord-books_app/pull/5/files#r792132956

pofkuma committed:
エラー発生時は固定のアラートメッセージを表示させる

https://github.com/pofkuma/fjord-books_app/pull/5/files#r792218473
https://github.com/pofkuma/fjord-books_app/pull/5/files#r792297476

pofkuma committed:
コントローラーで取得したインスタンスを利用

https://github.com/pofkuma/fjord-books_app/pull/5/files#r792299714

pofkuma committed:
i18n対応漏れ

pofkuma committed:
フォロー済みかどうかの判定処理をメソッド化

pofkuma committed:
フォロー一覧とフォロワー一覧のビューを共通化

https://github.com/pofkuma/fjord-books_app/pull/5/files#r792299179

betachelsea commented on app/models/user.rb:
```diff
@@ -5,4 +5,18 @@ class User < ApplicationRecord
+           dependent: :destroy, inverse_of: :following
+  has_many :followers, through: :follower_relationships, source: :follower
+
+  def following?(user_id:)
```
[IMO]

そのメソッドが動くには引数が必須で、しかも1つだけの場合は、キーワード引数とせずに普通に渡してしまうのもありかなと思いました。キーワード引数が有効なのはたくさん引数があって順番を把握するのが大変な時とか...かもしれないです。(個人によって感じ方は違いそうです 🙏   一意見まで！)

また、user_id をもらっていますが、どんな数字でも受け取れてしまうので、あえてUserのインスタンスをもらうようにしても良いかなと思いました。

```ruby
# 利用イメージ
me = User.find(params[:user_id])
other = User.find(params[:other_id])
me.following?(other)
```

pofkuma replied:
助言くださりありがとうございます！
そのような使い方にしたほうが良いと思いましたので、 dd727cc944cbbc93a2f481f16b3e9bfa36c18d0d にて修正しました。

betachelsea commented on app/views/users/shared/_lists.html.erb:
```diff
@@ -0,0 +1,33 @@
+<h1><%= t('.title') %></h1>
```
[NITS] この表記が `app/views/users/followings/index.html.erb`  `app/views/users/followers/index.html.erb` からpartialに移設されてきたため、一覧ページのタイトルがi18nの適用から漏れてしまっていますね。
見出しやリンクは元の `index.html.erb` に残しておいて partial化の対象とはしない、という方法がよくあると思います〜 👀 

```ruby
# app/views/users/followings/index.html.erb

<h1><%= t('.title') %></h1>

<%= render 'users/shared/lists', users: @followings %>

<%= link_to t('views.common.back'), user_path(@user) %>
```

partial化の単位の見極めは時と場合によるので難しいところではありますが、使い回したときに出し分けしたくなる可能性のある要素は含めないようにすると良いかなと思います。(例えば ... このpartialを使うと必ず「戻る」リンクがくっついてくる。それはAのページでは出したいけど、Bのページでは不要...というときは、「戻る」リンクはpartialには含めない。など。)

pofkuma replied:
i18nの部分、確認不足で申し訳ないです🙇‍♀️
提案してくださったやり方を適用させていただきました。-> 9a77417c72161765ba184902fc819c29f2162d18
どの部分をpartial化するか、今後慎重に見極めるようにしたいと思います！

betachelsea reviewed (approved):
確認OKです！ 細かい点だけ少しコメントしました 📝 
commitの粒度が綺麗で素晴らしいと思いました〜 ありがとうございます✨

pofkuma committed:
idをUserのインスタンスから取得

単なる数字を受け取るのではなく、確実にUserのidを受け取れるようにする

pofkuma committed:
見出しとリンクは共通化対象としない

ページごとに出し分けしそうな要素は、各々のページに持たせる
